x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ml;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Assertions;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * The version number associated with various ML features.
+ * <p>
+ * Prior to 8.10.0, the release {@link Version} was used everywhere. This class separates the ML config format version
+ * from the running node version.
+ * <p>
+ * Each ML config version constant has an id number, which for versions prior to 8.10.0 is the same as the release version
+ * for backwards compatibility. In 8.10.0 this is changed to an incrementing number, disconnected from the release version,
+ * starting at 10.
+ * <p>
+ * Each version constant has a unique id string. This is not actually used in the binary protocol, but is there to ensure
+ * each protocol version is only added to the source file once. This string needs to be unique (normally a UUID,
+ * but can be any other unique nonempty string).
+ * If two concurrent PRs add the same ML config version, the different unique ids cause a git conflict, ensuring the second PR to be merged
+ * must be updated with the next free version first. Without the unique id string, git will happily merge the two versions together,
+ * resulting in the same ML config version being used across multiple commits, causing problems when you try to upgrade between those two
+ * merged commits.
+ * <h2>Version compatibility</h2>
+ * The earliest compatible version is hardcoded in the {@link #MINIMUM_COMPATIBLE} field. This cannot be dynamically calculated
+ * from the major/minor versions of {@link Version}, because {@code MlConfigVersion} does not have separate major/minor version numbers.
+ * So the minimum compatible version is hard-coded as the ML config version used by the first version of the previous major release.
+ * {@link #MINIMUM_COMPATIBLE} should be updated appropriately whenever a major release happens.
```

Comment:
This part isn't true. We don't aggressively upgrade every single config when moving to a new major version. It's possible that there are some ML config documents in 8.x clusters that predate 7.0.